src/dune_pkg/solver_env.mli

```diff
@@ -1,84 +1,99 @@
-  end
-end
+      val set : t -> sys_var -> string -> t
 
-module Sys_var : sig
-  (** A system environment variable. Each of these variables may be assigned a
-      string value. If system environment variable has a value then that
-      variable binding will be used when evaluating opam dependency filters.
-      Unset variables are treated as wildcards. *)
-  type t =
-    [ `Arch
-    | `Os
-    | `Os_version
-    | `Os_distribution
-    | `Os_family
-    | `Opam_version
-    ]
+      val get : t -> sys_var -> string option
 
-  val to_string : t -> string
+      type union_error =
+        [ `Var_in_both_with_different_values of sys_var * string * string ]
 
-  val of_string_opt : string -> t option
+      (** Merge two environments returning an error if they both contain a
+          binding of the same variable to different values. *)
+      val union : t -> t -> (t, union_error) result
+    end
+  end
+end
 
-  module Bindings : sig
-    type sys_var := t
+module With_opt_build_config : sig
+  (** The default value for the "make" build config variable can't be computed
+      until after the workspace file has been parsed as it depends on the build
+      context. This types represents an environment where some build config
+      variables might be unset and can be converted into a full environment by
+      setting unset variables to their default values. *)
+  type t
 
-    (** A mapping from system environment variables to their values *)
-    type t
+  val default : t
 
-    val empty : t
+  val decode : t Dune_lang.Decoder.t
 
-    val set : t -> sys_var -> string -> t
+  val to_dyn : t -> Dyn.t
 
-    val get : t -> sys_var -> string option
+  val equal : t -> t -> bool
+end
 
-    type union_error =
-      [ `Var_in_both_with_different_values of sys_var * string * string ]
+(** A typed variable environment used by the dependency solver to evaluate
+    package filters. Opam packages can declare conditional dependencies on other
+    packages using a language made up of boolean operators and comparisons of
+    strings. Variables in this language can represent booleans and strings. *)
+type t
 
-    (** Merge two environments returning an error if they both contain a binding
-        of the same variable to different values. *)
-    val union : t -> t -> (t, union_error) result
-  end
-end
+(** Create a [t] from a [With_opt_build_config.t] filling in any unset fields
+    from the build config with default values. The [which_make] function is used
+    to look up the path of the `make` executable if the "make" build config
+    variable is unset. The [which_make] argument is ideally a function which
+    calls [Dune_rules.Expander.which_make] for the appropriate context. That
+    function can't be called directly from this module as it would create a
+    circular dependency. *)
+val with_default_values_for_unset_build_configs :
+     With_opt_build_config.t
+  -> which_make:(unit -> Path.t option Memo.t)
```

Comment:
Why does this argument need to be delayed?

Reply:
We only need to call it if there is no `make` variable configured.

Reply:
Should we make `%{make}` part of the solver environment at all? Selecting a package based on the concrete path of make sounds like a very bad idea to me. Should we just forbid it?

Reply:
Good point. I added them so that we could generate lockfiles with the build and install commands fully evaluated as a first step before supporting filters and substitutions in build and install commands but given the complexity added by these variables and the fact that packages _shouldn't_ be filtering based on `%{make}` or `%{jobs}` I'm inclined to remove them. Instead I'll generate lockfiles with unfiltered commands and convert variables like `%{make}` and `%{jobs}` into the appropriate pforms.